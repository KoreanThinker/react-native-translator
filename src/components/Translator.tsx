import * as React from 'react';
import {useCallback, useEffect, useMemo, useState} from 'react';
import _ from 'lodash';
import WebView, {WebViewMessageEvent} from 'react-native-webview';
import {LanguageCode, SourceLanguageCode, TranslatorType} from '..';
import translators from '../translators';
import {LOADING_MESSSAGE} from '../classes/translator';

export interface TranslatorProps<T extends TranslatorType> {
  from: SourceLanguageCode<T>;
  to: LanguageCode<T>;
  value: string;
  type?: T;
  onTranslated: (result: string) => void;
}

function Translator<T extends TranslatorType = 'google'>(
  props: TranslatorProps<T>,
) {
  const {from, to, value: _value, onTranslated, type = 'google'} = props;
  const [value, setValue] = useState('');
  const translator = useMemo(() => translators[type], [type]);
  const injectedJavascript = useMemo(
    () => translator.getInjectedJavascript(),
    [translator],
  );
  const userAgent = useMemo(() => translator.userAgent, [translator]);
  const uri = useMemo(
    () => (from && to ? translator.toUrl(from, to, value) : ''),
    [translator, from, to, value],
  );

  const onMessage = useCallback((event: WebViewMessageEvent) => {
    const result = event.nativeEvent.data;
    if (!result || result === LOADING_MESSSAGE || result === 'Enter a URL') {
      return;
    }
    onTranslated(result);
  }, []);

  // set value throttled
  const throttledSetValue = useMemo(() => _.debounce(setValue, 100), []);
  useEffect(() => {
    if (_value === '') {
      // clear value when input value is empty
      setValue('');
      onTranslated('');
      return;
    }
    throttledSetValue(_value);
  }, [_value]);

  return (
    <WebView
      style={{width: 0, height: 0}}
      injectedJavaScript={injectedJavascript}
      userAgent={userAgent}
      source={{uri}}
      onMessage={onMessage}
      cacheEnabled
    />
  );
}

export default Translator;
