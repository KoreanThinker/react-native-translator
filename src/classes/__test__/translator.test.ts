import Translator from '../translator';

const params = {
  lanaugeCodes: ['en', 'ko'],
  selector: '#result',
  toUrl: (from: string, to: string, value: string) =>
    `https://example.com/?sl=${from}&tl=${to}&text=${value}`,
};

test('constructor', () => {
  const translator = new Translator(params, {
    beforeTranslate: 'console.log("beforeTranslate")',
    userAgent: 'test user agent',
  });

  expect(translator.lanaugeCodes).toStrictEqual(['en', 'ko']);
  expect(translator.selector).toBe('#result');
  expect(translator.userAgent).toBe('test user agent');
  expect(translator.beforeTranslate).toBe('console.log("beforeTranslate")');
  expect(translator.toUrl('en', 'ko', 'hello')).toBe(
    'https://example.com/?sl=en&tl=ko&text=hello',
  );
});

test('getInjectedJavascript', () => {
  const translator = new Translator(params, {
    beforeTranslate: 'console.log("beforeTranslate")',
  });

  const injectedJavascript = translator.getInjectedJavascript();
  expect(injectedJavascript).toBe(`
      // if beforeTranslate is set, run it
      console.log("beforeTranslate")

      var selector = #result
      // Wait for the element to be loaded
      setInterval(() => {
        var result = document.querySelector(selector)
        if(result) {
          window.ReactNativeWebView.postMessage(result.innerText)
        }
        else {
          window.ReactNativeWebView.postMessage('@L@O@A@D@I@N@G@')
        }
      }, 200)
    `);
  expect(injectedJavascript).toContain('console.log("beforeTranslate")'); // beforeTranslate
  expect(injectedJavascript).toContain('#result'); // selector
});

test('getInjectedJavascript without beforeTranslate', () => {
  const translator = new Translator(params);
  const injectedJavascript = translator.getInjectedJavascript();
  expect(injectedJavascript).not.toContain('console.log("beforeTranslate")');
});
