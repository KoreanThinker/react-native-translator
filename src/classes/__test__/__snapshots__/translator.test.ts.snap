// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getInjectedJavascript 1`] = `
"
      // if beforeTranslate is set, run it
      console.log("beforeTranslate")

      var selector = #result
      // Wait for the element to be loaded
      setInterval(() => {
        var result = document.querySelector(selector)
        if(result) {
          window.ReactNativeWebView.postMessage(result.innerText)
        }
        else {
          window.ReactNativeWebView.postMessage('@L@O@A@D@I@N@G@')
        }
      }, 200)
    "
`;
